### PR TITLE
Fix hang caused by `steal` command with empty test queue

### LINKED ISSUE #884: --dist=worksteal + high core counts leads to hangs
I'm able to consistently reproduce with https://github.com/pyca/cryptography on an M1 MBP, which has 10 cores (8 performance, 2 efficiency).

Specifically `py.test -n 10 --dist worksteal --last-failed tests/x509/` consistently hangs.

It consistently works if I remove the `--dist worksteal` flag. Using lower core counts (e.g. `-n 8`) makes the hangs more intermittent, but the hangs still occur.

The hangs happen at the end of the test suite -- I believe after all the tests have run.

If there's any additional information I can provide, pleaese let me know.

### PR DESCRIPTION
**Pull request by amezin**
Fixes #884

### CODE REVIEW

File: changelog/884.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix hang caused by `steal` command with empty test queue (`worksteal` scheduler).
```
Comment (nicoddemus) [inline changelog/884.bugfix:1]:
```suggestion
Fixed hang using ``worksteal`` scheduler using a high number of CPU cores.
```

File: src/xdist/remote.py [hunk 2]
```diff
@@ -72,6 +73,12 @@ def __init__(self, config, channel):
     def _make_queue(self):
         return self.channel.gateway.execmodel.queue.Queue()
 
+    def _get_next_item_index(self):
```
Comment (nicoddemus) [inline src/xdist/remote.py:76]: It would be nice to add a docstring explaining why the retry is needed.

**Review by nicoddemus — APPROVED** [review]
Left a few minor suggestions, otherwise LGTM. :+1: 

Thanks for tackling this so quickly.

After this gets merged, feel free to prepare a new release.

File: changelog/884.bugfix (hunk 1 above)
Comment (nicoddemus) [reply changelog/884.bugfix:1]: The rationale here is that changelogs are read by end-users which are not usually familiarized with the internal, so better use a language closer to the usage of the plugin.

File: src/xdist/remote.py (hunk 2 above)
Comment (amezin) [reply src/xdist/remote.py:76]: Will `RETRY_MARK` -> `QUEUE_CHANGED_MARK` make the code more self-explanatory?

File: changelog/884.bugfix (hunk 1 above)
Comment (amezin) [reply changelog/884.bugfix:1]: But "high number of CPU cores" has nothing to do with the issue. I can reproduce it with `-n 3`

Comment (nicoddemus) [reply changelog/884.bugfix]: Ahh OK, feel free to reword it. :+1:

File: src/xdist/remote.py (hunk 2 above)
Comment (nicoddemus) [reply src/xdist/remote.py:76]: Ahh I think so. I would still add a docstring describing the issue however.

**Comment by reaperhulk** [comment]
Verified this stops the hanging we were seeing in the `pyca/cryptography` test suite. Thanks @amezin!

**Review by nicoddemus — APPROVED** [review]

**Pull request merged**